proposals/032-deprecate-unintended-apis.md

```diff
@@ -0,0 +1,221 @@
+SP #032: Deprecate Public APIs That Shouldn't Exist
+===================================================
+
+We propose to deprecate a large number of functions, enumerants, etc. from the Slang public API, and command-line options for `slangc`, in cases where any of the following is true:
+
+* the features were never intended for public use
+
+* the features represent design choices we regret
+
+Status
+------
+
+Status: Implementation In-Progress
+
+Implementation: -
+
+Author: Theresa Foley (@tangent-vector)
+
+Reviewer: -
+
+Background
+----------
+
+This proposal touches on two main categories of features of the Slang compiler codebase:
+
+### Features That Were Never Intended for Public Use
+
+Slang's build process involves some "boostrapping" steps, including needing to invoke the Slang compiler on the source code for the builtin modules, so that the compiled binary for the core module can be embedded into the Slang compiler itself.
+Unfortunately, at the time that the bootstrapping process was implemented, the hooks that were required to implement it got declared in the public `slang.h` header, thus enabling others to use those hooks, even if they were never intended to be part of the supported API of the compiler. The same applies to several command-line options for `slangc` (which also surface as enumerants in `slang.h`) that are only intended to be used during the bootstrapping process.
+
+In addition to the bootstrapping case, there are several parts of the API in `slang.h` or options for `slangc` that only exist to support the `slang-test` tool and the approach we use for testing Slang.
+An important example is the support for "pass-through" compilation, where `slangc` doesn't actually try to compile a file and instead hands it off to a downstream compiler more-or-less untouched; the cost of supporting this functionality has risen greatly over time, and it only exists to support some of our earliest test cases.
+
+### Features That We Regret
+
+Some things got put into the `slang.h` API that were over-designed from the start, and never saw enough uptake among developers to justify the cost of supporting all the bells and whistles.
+Examples include:
+
+* `ISlangCastable` and `ISlangClonable`
+
+* `ISlangFileSystemExt`
+
+* `ISlangWriter`
+
+* The ability for users to override the "prelude" code for particular target languages and/or downstream compilers
```

Comment:
We are currently using a custom build of Slang that uses our own C++ prelude in addition to some other changes. This feature could be a potentially useful way for us to use the official build of Slang.

Our intention is to contribute our changes back, but some of our prelude changes rely on LLVM compiler plugins so I'm not sure if all our prelude changes will make it in.